File: app/java/MoviesApp/app/src/main/java/com/example/moviesapp/data/movies/impl/MoviesDtoConverter.java
package com.example.moviesapp.data.movies.impl;

import com.example.moviesapp.data.movies.dto.MoviesListDto;
import com.example.moviesapp.data.movies.model.Movie;
import com.example.moviesapp.data.movies.model.MovieStatistics;

import java.util.List;
import java.util.function.Function;
import java.util.stream.Collectors;

/**
 * Converter that converts the movies dto to a list of movies.
 */
public class MoviesDtoConverter implements Function<MoviesListDto, List<Movie>> {

    @Override
    public List<Movie> apply(MoviesListDto moviesDto) {
        return moviesDto.movies
                .stream()
                .map(movieDto -> new Movie(
                        movieDto.title,
                        movieDto.id,
                        movieDto.imageUrl,
                        movieDto.statistics != null ? new MovieStatistics(
                                movieDto.statistics.score,
                                movieDto.statistics.percentViewed
                        ) : new MovieStatistics(
                                null,
                                null
                        )
                )).collect(Collectors.toList());
    }
}

File: app/android/kotlin/MoviesApp/app/src/main/java/com/example/moviesapp/data/movies/impl/MoviesDtoConverter.kt
package com.example.moviesapp.data.movies.impl

import com.example.moviesapp.data.movies.dto.MoviesListDto
import com.example.moviesapp.data.movies.model.Movie
import com.example.moviesapp.data.movies.model.MovieStatistics

/**
 * Converter that converts the movies dto to a list of movies.
 */
internal class MoviesDtoConverter: (MoviesListDto) -> List<Movie> {

    override fun invoke(input: MoviesListDto) = input.movies.map {
        with(it) {
            Movie(
                    title = title,
                    id = id,
                    imageUrl = imageUrl,
                    statistics = with(it.statistics) {
                        MovieStatistics(
                                score = score.toBigDecimal(),
                                percentViewed = percentViewed.toBigDecimal()
                        )
                    }
            )
        }
    }
}